### DISCUSSION TITLE
None guard with function that may not return

### DISCUSSION THREAD

**Discussion by diego-pm** [microsoft/pyright#4033] [Q&A] [answered]
I am looking to None guard a variable with a function that raises an exception when the variable is None. I have tried to annotate the function with overloading, and specifying that the function does not return when the variable is None. See this simplified example:

```python
@overload
def check_not_none(v: str) -> None: ...
@overload
def check_not_none(v: None) -> NoReturn: ...
def check_not_none(v: str | None) -> None:
    if v is None:
        raise ValueError("Value must not be None")
    
def f(x: str | None) -> str:
    check_not_none(x)
    return x.upper()
```

However, pyright still gives the error that `upper` is not a known member of `None` in function `f`. Is there a way I could solve this?

**Comment by erictraut** [comment, marked as answer]
Unguarded function calls do not apply type narrowing to arguments passed to them, so the approach you're using above won't work.

The feature you're looking for here is what is called a "type assert" in languages like TypeScript. It doesn't currently exist in the Python type system. I've proposed such a feature as an extension to PEP 647, but this proposal hasn't gone beyond early discussions. For details refer to [this thread](https://github.com/python/typing/discussions/1013) and search for `TypeAssert`.

In the meantime, you will need to use the built-in assert to handle this case.

```python
def f(x: str | None) -> str:
    assert x is not None
    return x.upper()
```

Or you can use a function that returns a "cleaned" value.

```python
from typing import NoReturn, TypeVar, overload

T = TypeVar("T")

def check_not_none(v: T | None) -> T:
    if v is None:
        raise ValueError("Value must not be None")
    return v
    
def f(x: str | None) -> str:
    x = check_not_none(x)
    return x.upper()
```

**Reply by diego-pm** [reply]
Thank you for the explanation and solutions. TypeAssert seems a good idea, hopefully one day it included in python.

**Reply by diego-pm** [reply]
Could it be possible that pyright applies type narrowing when using an overload with NoReturn as the in the example I was giving?

**Reply by erictraut** [reply]
That would a very different type narrowing mechanism than what's in place currently in pyright and in other type checkers like mypy. It would also be prohibitively expensive for pyright at code analysis time because it would require type evaluation of every argument expression that includes a target symbol. So no, that's not something we would implement.

**Reply by diego-pm** [reply]
Ok, thanks for the clarification. Also, what would be the best approach, in general, to tell the type checker that I am sure of the type of a variable? Should I use `cast`?

**Reply by erictraut** [reply]
I use `cast` as a last resort because it eliminates all type checking and therefore leads to fragile code. The techniques I mention above are preferable (using an `assert` statement or a function that returns a value after verifying its type).
